documentation/smg/16_serialiser.md

```diff
@@ -0,0 +1,229 @@
+
+- R - Rank
+- T - Type
+
+### Type Tags
+Tag| Meaning
+---|----------------------
+  0| LOGICAL
+  1| CHAR
+  2| MATLAB_STRING
+  3| DOUBLE
+  4| SINGLE
+  5| INT8
+  6| UINT8
+  7| INT16
+  8| UINT16
+  9| INT32
+ 10| UINT32
+ 11| INT64
+ 12| UINT64
+ 13| COMPLEX_DOUBLE
+ 14| COMPLEX_SINGLE
+ 15| COMPLEX_INT8
+ 16| COMPLEX_UINT8
+ 17| COMPLEX_INT16
+ 18| COMPLEX_UINT16
+ 19| COMPLEX_INT32
+ 20| COMPLEX_UINT32
+ 21| COMPLEX_INT64
+ 22| COMPLEX_UINT64
+ 23| CELL
+ 24| STRUCT
+ 25| FUNCTION_HANDLE
+ 26| VALUE_OBJECT
+ 27| HANDLE_OBJECT_REF
+ 28| ENUM
+ 29| SPARSE_LOGICAL
+ 30| SPARSE_DOUBLE
+ 31| SPARSE_COMPLEX_DOUBLE
+
+### Serialisation formats in detail
+
+#### Simple data
+Simple data (type tags 0-12) are serialised as:
+
+_Rank_ |     tag    | dim1 | dim2 | ... | Data
+-------|------------|------|------|:---:|------
+_NULL_ | 32 + type  | 0
+_0_    |    type    |      |      |     | Data
+_1_    | 32 + type  | nElem|      |     | Data
+_2_    | 64 + type  | dim1 | dim2 |     | Data
+_N_    | N<<5 + type| dim1 | dim2 | ... | Data
+
+#### Complex Data
+Complex numerical data (type tags 13-22) are serialised as:
+
+_Rank_ |     tag    | dim1 | dim2 | ... |    Data   |   ...
+-------|------------|------|------|:---:|-----------|----------
+_NULL_ | 32 + type  | 0    |      |     |           |
+_0_    |    type    |      |      |     | Real Data | Imag Data
+_1_    | 32 + type  | nElem|      |     | Real Data | Imag Data
+_2_    | 64 + type  | dim1 | dim2 |     | Real Data | Imag Data
+_N_    | N<<5 + type| dim1 | dim2 | ... | Real Data | Imag Data
+
+#### Sparse data
+Sparse numeric data (type tags 29-31) are serialised as:
+
+_Rank_ |   Type  |     tag  | dim1 | dim2 | Data | ... |    ...    | ...
+-------|---------|----------|------|------|:----:|:---:|-----------|------
+_2_    | Real    | 64 + type| dim1 | dim2 |   i  |  j  | Data      |
+_2_    | Bool    | 64 + type| dim1 | dim2 |   i  |  j  | Data      |
+_2_    | Complex | 64 + type| dim1 | dim2 |   i  |  j  | Real Data | Imag Data
+
+Based on `sparse(i,j,data)` [MATLAB documentation][sparse]
+
+__NB.__ only rank-2 sparse arrays are permitted in MATLAB
+
+__NB.__ Due to storing dimensions as `uint32` data, rather than `double`, sizes of sparse arrays are limited to (2^32)-1
+
+#### Struct data
+Structured tree data (type tag 24) uses the standard header format:
+
+_Rank_ |     tag    | dim1 | dim2 | ... | Data
+-------|------------|------|------|:---:|------
+_NULL_ | 32 + type  | 0
+_0_    |    type    |      |      |     | Data
+_1_    | 32 + type  | nElem|      |     | Data
+_2_    | 64 + type  | dim1 | dim2 |     | Data
+_N_    | N<<5 + type| dim1 | dim2 | ... | Data
+
+
+Where data is formatted as follows:
+
+nFields | length(fieldName1) | length(fieldName2) | ... | fieldName1 | fieldName2 | ... | struct2cell(data)
+--------|--------------------|--------------------|:---:|------------|------------|:---:|------------------
+
+nFields and fieldName lengths are serialised as `uint32`
+
```

Comment:
Again, discussing the data format - which may be good, but should be consistent.